pkg/virt-handler/vm.go

```diff
@@ -408,115 +529,199 @@ func (d *VMHandlerDispatch) injectDiskAuth(vm *v1.VirtualMachine) (*v1.VirtualMa
 
 	// If we don't have the VM in the cache, it could be that it is currently migrating to us
-	result := d.restClient.Get().Name(vmName).Resource("virtualmachines").Namespace(namespace).Do()
-	if result.Error() == nil {
+	fetchedVM, err := d.clientset.VM(namespace).Get(vmName, metav1.GetOptions{})
+	if err == nil {
 		// So the VM still seems to exist
-		fetchedVM, err := result.Get()
-		if err != nil {
-			return false, err
-		}
-		if fetchedVM.(*v1.VirtualMachine).Status.MigrationNodeName == d.host {
+
+		if fetchedVM.Status.MigrationNodeName == d.host {
 			return true, nil
 		}
-	} else if !errors.IsNotFound(result.Error()) {
+	} else if !errors.IsNotFound(err) {
 		// Something went wrong, let's try again later
-		return false, result.Error()
+		return false, err
 	}
 
 	// VM object was not found.
 	return false, nil
 }
 
 func isWorthSyncing(vm *v1.VirtualMachine) bool {
-	return vm.Status.Phase != v1.Succeeded && vm.Status.Phase != v1.Failed
+	return !vm.IsFinal()
+}
+
+func (d *VirtualMachineController) checkFailure(vm *v1.VirtualMachine, syncErr error, reason string) (changed bool) {
+	if syncErr != nil && !d.hasCondition(vm, v1.VirtualMachineSynchronized) {
+		vm.Status.Conditions = append(vm.Status.Conditions, v1.VMCondition{
+			Type:               v1.VirtualMachineSynchronized,
+			Reason:             reason,
+			Message:            syncErr.Error(),
+			LastTransitionTime: metav1.Now(),
+			Status:             k8sv1.ConditionFalse,
+		})
+		return true
+	} else if syncErr == nil && d.hasCondition(vm, v1.VirtualMachineSynchronized) {
+		d.removeCondition(vm, v1.VirtualMachineSynchronized)
+		return true
+	}
+	return false
+}
+
+func (d *VirtualMachineController) hasCondition(vm *v1.VirtualMachine, cond v1.VirtualMachineConditionType) bool {
+	for _, c := range vm.Status.Conditions {
+		if c.Type == cond {
+			return true
+		}
+	}
+	return false
+}
+
+func (d *VirtualMachineController) removeCondition(vm *v1.VirtualMachine, cond v1.VirtualMachineConditionType) {
+	conds := []v1.VMCondition{}
+	for _, c := range vm.Status.Conditions {
+		if c.Type == cond {
+			continue
+		}
+		conds = append(conds, c)
+	}
+	vm.Status.Conditions = conds
+}
+
+func (d *VirtualMachineController) setVmPhaseForStatusReason(domain *api.Domain, vm *v1.VirtualMachine) {
+	vm.Status.Phase = d.calculateVmPhaseForStatusReason(domain, vm)
+}
+func (d *VirtualMachineController) calculateVmPhaseForStatusReason(domain *api.Domain, vm *v1.VirtualMachine) v1.VMPhase {
+
+	if domain == nil {
+		if !vm.IsRunning() {
+			return v1.Scheduled
+		} else {
+			// That is unexpected. We should not be able to delete a VM before we stop it.
+			// However, if someone directly interacts with libvirt it is possible
+			return v1.Failed
+		}
+	} else {
+		switch domain.Status.Status {
+		case api.Shutoff, api.Crashed:
+			switch domain.Status.Reason {
+			case api.ReasonCrashed, api.ReasonPanicked:
```

Comment:
Did you mean to leave out "ReasonFailed" from this case statement?

Also, We might want to reconsider how this switch case is formatted. Maybe success should be a whitelist of Reasons, migration should be ignored, and the failed case could be the default?  My concern is that without a default case, if libvirt adds another reason to the enum it would be possible that in some edge case our VMs might never move to a finalized state after being shutdown.

Reply:
I agree that this needs more love. I think we need to explicitly support all combinations, which can happen right now. If we see new (and potentially unsupported) state or reasons, I think we should should report that, and not guess. E.g. an unknown state, or not touching the state, but exposing the issue via a condition.

I can add a follow up PR.